core/templates/dev/head/expressions/typeParserSpec.js

```diff
@@ -0,0 +1,122 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+describe('Expression type parser service', function() {
+  beforeEach(module('oppia'));
+
+  var ees = null;
+  var eps = null;
+  beforeEach(inject(function($injector) {
+    ees = $injector.get('expressionTypeParserService');
+    eps = $injector.get('expressionParserService');
+    ests = $injector.get('expressionSyntaxTreeService');
+  }));
+
+  var ENVS = [
+    {
+      numZero: 'Real',
+      boolTrue: 'UnicodeString',
+      strXYZ: 'UnicodeString',
+      num100_001: 'Real',
+      boolFalse: 'UnicodeString',
+      strNull: 'UnicodeString'
+    }
+  ];
+
+  it('should determine the correct types for the expressions', function() {
+    [
+      ['2', 'Real'],
+      ['numZero', 'Real'],
+      ['boolTrue', 'UnicodeString'],
+      ['+10', 'Real'],
+      ['2   + 10', 'Real'],
+      ['num100_001   + numZero', 'Real'],
+      ['20 - num100_001', 'Real'],
+      ['0x100 - 256', 'Real'],
+      ['!strNull', 'UnicodeString'],
+      ['1 - 2 * 3', 'Real'],
+      ['num100_001 / 0.1', 'Real'],
+      ['floor((numZero + num100_001)/2)', 'Real'],
+      ['23 % 5', 'Real'],
+      ['1 <= numZero || 1 >= numZero', 'UnicodeString'],
+      ['100 < num100_001 && 1 > num100_001', 'UnicodeString'],
+      ['boolTrue == boolFalse', 'UnicodeString'],
+      ['strNull != strXYZ', 'UnicodeString'],
+      ['if boolFalse then 8 else numZero', 'Real'],
+      ['if boolFalse then 8 else strXYZ', ests.ExprWrongArgTypeError],
```

Comment:
Why yes. And now I have a few. Added an additional test for an error condition that wasn't tested before.